server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3083,21 +3084,44 @@ assert getActiveOperationsCount() == OPERATIONS_BLOCKED
         final TranslogStats translogStats = translogStats();
         // flush to make sure the latest commit, which will be opened by the read-only engine, includes all operations.
         flush(new FlushRequest().waitIfOngoing(true));
-        synchronized (mutex) {
-            verifyNotClosed();
-            // we must create a new engine under mutex (see IndexShard#snapshotStoreMetadata).
-            final Engine readOnlyEngine = new ReadOnlyEngine(newEngineConfig(), seqNoStats, translogStats, false, Function.identity());
-            IOUtils.close(currentEngineReference.getAndSet(readOnlyEngine));
-        }
 
         Engine newEngine = null;
+        SetOnce<Engine> newEngineReference = new SetOnce<>();
         try {
             final long globalCheckpoint = getGlobalCheckpoint();
             synchronized (mutex) {
-                assert currentEngineReference.get() instanceof ReadOnlyEngine : "another write engine is running";
                 verifyNotClosed();
                 // we must create a new engine under mutex (see IndexShard#snapshotStoreMetadata).
+                final Engine readOnlyEngine = new ReadOnlyEngine(newEngineConfig(), seqNoStats, translogStats, false, Function.identity()) {
+                    @Override
+                    public IndexCommitRef acquireLastIndexCommit(boolean flushFirst) {
+                        synchronized (mutex) {
+                            return newEngineReference.get().acquireLastIndexCommit(flushFirst);
+                        }
+                    }
+
+                    @Override
+                    public IndexCommitRef acquireSafeIndexCommit() {
+                        synchronized (mutex) {
+                            return newEngineReference.get().acquireSafeIndexCommit();
+                        }
+                    }
+
+                    @Override
+                    public void close() throws IOException {
+                        assert Thread.holdsLock(mutex);
+
+                        Engine newEngine = newEngineReference.get();
```

Comment:
With this change, I think we can remove the try/finally block.

Reply:
Fixed in 6888b6612c3f2b880afaffdd91f4865ec6433ab5